### PR TITLE
Allow to use cover from sub directory

### LINKED ISSUE #30: Broken link when cover image is in sub-gallery (w/ proposed solution)
I have a section that contains subsections.
For one section (`usa2017`) I would like to use a cover image from one of the sub directory.
I use this `settings.yaml`

```
title: USA 2017
sub_title: Road trip on the west coast
date: 2017-07-31
cover: ./Disney/3F6A0526.jpeg
```

~~This fails at build time with error~~
```
(usa2017/Disney/3F6A0526.jpeg) Failed to save "build/usa2017/Disney/3F6A0526-1188218354-x900.jpeg". This usually happens when progressive is set to True and quality set to a too high value (> 90). Please lower quality or disable progressive, globally or for "usa2017/Disney/3F6A0526.jpeg". As a work-around, increase buffer size. This might result in side-effects.
The original error is "[Errno 2] No such file or directory: 'build/usa2017/Disney/3F6A0526-1188218354-x900.jpeg'"
```

UPDATE : there is no build error but the generated web page has an incorrect link.  Check comment https://github.com/recitale/recitale/issues/30#issuecomment-1705837950 below for details.

Here is my directory layout
```
usa2017
├── Death Valley
├── Disney
├── Grand Canyon
...
```

I have a fix working on my machine. It involves some changes in `Images.py` to have `Thumbnail` return the full path and changes in the template HTML.

I can send a PR to discuss the approach of the fix, but I want to wait that https://github.com/recitale/recitale/pull/29 is merged first.

### PR DESCRIPTION
**Pull request by sebsto**
This is a proposed solution to fix https://github.com/recitale/recitale/issues/30

Goal : avoid unneeded copy (duplication) of images by allowing a gallery to refer to a cover image in a sub gallery (sub directory)

This approaches modify images.py to ensure the Thumbnail's path contains the full path and not only the file name.

Happy to discuss alternatives if you think there are better ways to solve this problem.

### CODE REVIEW

**Comment by QSchulz** [comment]
Hi Sébastien, thanks for the PR.

Aside from not being to reproduce the issue in #30, I actually have a few initial thoughts on this PR.

First, there are multiple fixes in there (thanks for catching them!) so separate commits for each would be appreciated. They all deserve to be recognized as what they are, unrelated bugs that need to be fixed :)

I could identify (as unrelated to this MR):
- extending .gitignore to add VScode
- Home vs Back button (though... should we maybe fix it to return to home instead? no strong feeling about this one though)
- missing closing curly bracket in `recitale/themes/light/templates/index.html` for the video section

All of the above in a separate commit each, in a different merge request (or with this one but I anticipate it may take a few versions/discussions to merge the rest, see following paragraphs) will be happily merged as soon as it is made available and passes the GitHub CI.

As reported in #30, I am not able to reproduce the build issue which I would assume would be fixed with your change in https://github.com/recitale/recitale/pull/34/files#diff-7a1af9e3b56ba937ff0a1bdebc694c4defbbbf9d584e7d474c9cbfa0bd1cc982 (recitale/recitale.py), but it'd make sense on the practical level. I would recommend to use pathlib instead though:
```
filepath.parent.mkdir(parents=True, exist_ok=True)
```
which renders unnecessary the check on directory existence. Please provide a reproducer so I can make sure we're fixing everything (we need to think about video and audio too :) ).

As for the thing we for sure need to fix, specifically the broken links for gallery covers coming from subgalleries, I would very much like to avoid having to pass too much information to thumbnail/copy methods. We already should be giving enough info when creating the BaseImage thanks to the ImageFactory (called when doing an Image.get in templates). The templating should be extremely simple and straightforward to write, and concentrating all the difficult things in the ImageFactory makes sense to me. I have put some thoughts into this but need to take a bit more time to explain all the corner cases we need to cover. Right now, my gut feeling is that we could probably write some kind of wrapper around BaseImage and return that, which will wrap the BaseImage.thumbnail and .copy method to add the original path. This wouldn't actually change any of the templates.

```
class TemplateImage:
    def __init__(self, baseimg, imgname):
        self.__baseimg = baseimg
        self.__imgname = imgname

    def thumbnail(self, size):
        filename = self.__baseimg.thumbnail(size)
        return self.__imgname.with_name(filename)
    [...]

# TODO: add support for looking into parent directories (name: ../other_gallery/pic.jpg)
class ImageFactory:
    base_imgs = dict()
    global_options = dict()

    @classmethod
    def get(cls, path, image):
        if not isinstance(image, dict):
            image = {"name": image}

        if "name" not in image:
            logger.error(
                'At least one image in "%s" does not have a `name` property, please add the '
                "filename of the image to a `name` property.",
                path + "/settings.yaml",
            )
            sys.exit(1)

        im = image.copy()
        imgname = im["name"]
        # To resolve paths with .. in them, we need to resolve the path first and then
        # find the relative path to the source (current) directory.
        im["name"] = Path(path).joinpath(im["name"]).resolve().relative_to(Path.cwd())
        img = BaseImage(im, cls.global_options)
        img = cls.base_imgs.setdefault(img.filepath / str(img.chksum_opt), img)
        return TemplateImage(img, imgname)
```

Code not tested, just verbatim from my brain.

The goal being to replace the original filename passed in the settings.yaml with the thumbnail filename while keeping the rest of the path that precedes it in settings.yaml.

The complexity of the final implementation will boil down to how far we want to go with that, because if we want to optimize the cache, that's going to require some thinking and changing the data structures I believe.

So in short for the things to not forget about:
- video support
- audio support
- light theme as principal theme
- light theme as secondary theme
- material theme
- cache should stay coherent
- update documentation
- update unit tests

Finally, we'll need to cover any new code in unit tests and the automated runs in GitHub CI will have to pass (you can run the tests locally with `tox r`). The documentation should probably be updated to explicit this is supported.

**Comment by sebsto** [comment]
Thank you for your review - I will proceed step by step

First, the three unrelated changes have now each their own PR

https://github.com/recitale/recitale/pull/37
https://github.com/recitale/recitale/pull/39
https://github.com/recitale/recitale/pull/41

**Comment by sebsto** [comment]
> As reported in https://github.com/recitale/recitale/issues/30, I am not able to reproduce the build issue which I would assume would be fixed with your change in https://github.com/recitale/recitale/pull/34/files#diff-7a1af9e3b56ba937ff0a1bdebc694c4defbbbf9d584e7d474c9cbfa0bd1cc982 (recitale/recitale.py), but it'd make sense on the practical level. I would recommend to use pathlib instead though:
>
> filepath.parent.mkdir(parents=True, exist_ok=True)
>
> which renders unnecessary the check on directory existence. Please provide a reproducer so I can make sure we're fixing everything (we need to think about video and audio too :) ).

fixed in https://github.com/recitale/recitale/pull/34/commits/99e380bc3381da48793a9f723f815227042fdee1

Thanks

**Comment by sebsto** [comment]
Thank you for sharing your thoughts about the fix for the broken link. I will need to digest this and will come back with an update.

You're right about adding test cases for new code. I haven't look at the project test structure and logic.  Will do.
